### PR TITLE
Turn on userns-remap=dockerremap

### LINKED ISSUE #1134: `--userns-remap` feature not supported?
When I add `--userns-remap=defaut` to `EXTRA_ARGS` in `/var/lib/boot2docker/profile`, then `docker-machine start ...` reports:

`Unable to verify the Docker daemon is listening: Maximum number of retries (10) exceeded`

When I stop docker with `sudo /etc/init.d/docker stop` and then run the command manually, I get a fatal error:

```
docker@default:~$ sudo /usr/local/bin/docker daemon -D -g /var/lib/docker -H unix:// -H tcp://0.0.0.0:2376 --label provider=vmwarefusion --tlsverify --tlscacer
t=/var/lib/boot2docker/ca.pem --tlscert=/var/lib/boot2docker/server.pem --tlskey=/var/lib/boot2docker/server-key.pem -s aufs --userns-remap=default
UEBU[0000] Trusting certs with subjects: [010
tailee] 
DEBU[0000] docker group found. gid: 100                 
DEBU[0000] Server created for HTTP on unix (/var/run/docker.sock) 
DEBU[0000] Server created for HTTP on tcp (0.0.0.0:2376) 
FATA[0000] Error starting daemon: Error during "dockremap" user creation: Error adding group "dockremap": Cannot add group; no groupadd/addgroup binary found 
```

But there is an `addgroup` binary:

```
docker@default:~$ which addgroup
/usr/sbin/addgroup
```

If I add my own group and user manually, and try to use those, I get a different error:

```
docker@default:~$ sudo addgroup -g 20 osx_staff
docker@default:~$ sudo adduser -G osx_staff -SDHu 501 osx_tailee
docker@default:~$ sudo /usr/local/bin/docker daemon -D -g /var/lib/docker -H unix:// -H tcp://0.0.0.0:2376 --label provider=vmwarefusion --tlsverify --tlscacer
t=/var/lib/boot2docker/ca.pem --tlscert=/var/lib/boot2docker/server.pem --tlskey=/var/lib/boot2docker/server-key.pem -s aufs --userns-remap=501:20
UEBU[0000] Trusting certs with subjects: [010
tailee] 
DEBU[0000] docker group found. gid: 100                 
DEBU[0000] Server created for HTTP on unix (/var/run/docker.sock) 
DEBU[0000] Server created for HTTP on tcp (0.0.0.0:2376) 
INFO[0000] User namespaces: ID ranges will be mapped to subuid/subgid ranges of: osx_tailee:osx_staff 
FATA[0000] Error starting daemon: Can't create ID mappings: open /etc/subuid: no such file or directory 
```

I am not a linux guru. I tried reading the docker docs and linux man page for `subuid`, and came away less than enlightened.

What do I need to do to start using the new user namespace feature of docker 1.10 on OS X with `docker-machine` and `boot2docker`? I just want files written to volumes I share from my mac into a container to have the same owner and group as on my OS X machine (uid 501, gid 20).

### PR DESCRIPTION
**Pull request by SvenDowideit**
Signed-off-by: Sven Dowideit <SvenDowideit@home.org.au>

This should work - waiting for the iso to build locally

closes #1134 - enables --userns-remap by default.

### CODE REVIEW

**Comment by SvenDowideit** [comment]
mmmm, interesting possible side effect.

1 changing between userns-remap on and off seems to invalidate the build cache
2 I'm re-running atm, but it might be that turning userns-remap on will mean we can't build the b2d iso?

```
Step 46 : RUN cd /git &&     GIT_BRANCH=$(git rev-parse --abbrev-ref HEAD) &&     GITSHA1=$(git rev-parse --short HEAD) &&     DATE=$(date) &&     echo "${GIT_BRANCH} : ${GITSHA1} - ${DATE}" > $ROOTFS/etc/boot2docker
 ---> Using cache
 ---> b9d0b118a265
Step 47 : RUN cd $ROOTFS && zcat /tcl_rootfs.gz | cpio -f -i -H newc -d --no-absolute-filenames
 ---> Running in 4043cfc8719b
cpio: dev/tty6: Cannot mknod: Operation not permitted
cpio: dev/hdb1: Cannot mknod: Operation not permitted
cpio: dev/fd0: Cannot mknod: Operation not permitted
cpio: dev/ram6: Cannot mknod: Operation not permitted
```

**Comment by tianon** [comment]
See also the discussion in #1134 -- I'm definitely very much +1 to making sure that `EXTRA_ARGS='--userns-remap default'` works out of the box, but I'm not keen on turning it on by default unless/until Docker upstream does.

**Comment by tianon** [comment]
(which would mean replacing that username/group with `dockremap`, since that's what `--userns-remap default` expects / tries to create)

**Comment by tianon** [comment]
Also, couldn't we bake this user/group into the image instead of creating them at boot time?

**Comment by SvenDowideit** [comment]
@tianon yeah, we can - i didn't because we didn't with the `docker` user and group

I'm basically just following what we've been doing for ... er 2.5 years? 

ok, I'll push a version with it off by default

**Comment by SvenDowideit** [comment]
seriously? `dockremap`? ffs, i was reading way too fast - if only we had someone who could spell words.

**Comment by SvenDowideit** [comment]
ok, building and testing, and then I'll add some FAQ/doc

**Comment by SvenDowideit** [comment]
for FAQ:

When I start the daemon without --userns-remap, stop it, and then start with --userns-remap, it fails to start with 

```
time="2016-02-25T04:06:12.553255973Z" level=fatal msg="Error starting daemon: Couldn't create Tag store repositories: EOF"
```

`rm -r /var/lib/docker` fixes it.

**Comment by SvenDowideit** [comment]
```
root@test:/home/docker# rm -r /var/lib/docker/
root@test:/home/docker# /etc/init.d/docker start
Need TLS certs for test,127.0.0.1,10.0.2.15,192.168.99.101
-------------------
root@test:/home/docker# docker info
Containers: 0
 Running: 0
 Paused: 0
 Stopped: 0
Images: 0
Server Version: 1.10.2
Storage Driver: aufs
 Root Dir: /var/lib/docker/100000.100000/aufs
 Backing Filesystem: tmpfs
 Dirs: 0
 Dirperm1 Supported: true
Execution Driver: native-0.2
Logging Driver: json-file
Plugins:
 Volume: local
 Network: null host bridge
Kernel Version: 4.1.18-boot2docker
Operating System: Boot2Docker 1.10.2 (TCL 6.4.1); ttt : 930d115 - Thu Feb 25 03:54:25 UTC 2016
OSType: linux
Architecture: x86_64
CPUs: 1
Total Memory: 996.1 MiB
Name: test
ID: LTUC:SGH6:XNPR:MZ3T:3VIP:XYUR:RIF7:C33V:Q3EA:ZUDP:MUZM:KSVN
Debug mode (server): true
 File Descriptors: 10
 Goroutines: 21
 System Time: 2016-02-25T04:08:30.540681716Z
 EventsListeners: 0
 Init SHA1:
 Init Path: /usr/local/bin/docker
 Docker Root Dir: /var/lib/docker/100000.100000
Labels:
 provider=virtualbox
root@test:/home/docker# docker ps
CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS              PORTS               NAMES
root@test:/home/docker# docker run -d -it alpine top
Unable to find image 'alpine:latest' locally
latest: Pulling from library/alpine
ee54741ab35b: Pull complete
Digest: sha256:24a36bbc059b1345b7e8be0df20f1b23caa3602e85d42fff7ecd9d0bd255de56
Status: Downloaded newer image for alpine:latest
ca91ec41de82d5515046c7fb101b234671c5800766c65768ba296035fe7c314e
root@test:/home/docker# ps aux | grep top
100000    1811  0.6  0.0   1512     4 pts/1    Ss+  04:09   0:00 top
root      1818  0.0  0.0   9768   956 pts/0    S+   04:09   0:00 grep top
```

so \o/

**Comment by tianon** [comment]
The reason the Docker group bit is done during boot is specifically because that's handled by TCL itself -- _we_ don't ever create the Docker user, but we have to create the Docker group for it to have access to the socket (and which user is created can be modified via the kernel command line, which is the reason for the `if` there).  This is why I think we should pre-bake this user and their permissions during build instead (since it'll work the same but be already baked-in).

**Comment by tianon** [comment]
Other than that, this looks great. :smile: :+1:

**Comment by SvenDowideit** [comment]
mmm, i'll re-test this tomorrow - it might be that restarting causes it to go boom :(.

and idk who's to blame yet.

**Comment by SvenDowideit** [comment]
after one or three reboots, I start getting a failure to start the docker daemon:

```
msg="Error starting daemon: Couldn't create Tag store repositories: EOF"
```

it doesn't look like the uid/gid are changing. 

@estesp any quick advice on where to push?

File: doc/FAQ.md [hunk 1]
```diff
@@ -147,7 +147,24 @@ Or, if you need to tell the Docker daemon to use a specific DNS server, add the
...
+
+Boot2Docker can support enabling `--userns-remap` on the Docker daemon, but it is not enabled by default.
+
+To enable User namespacing, add `--userns-remap=default` to the `EXTRA_ARGS` in the `/var/lib/boot2docker/profile` file.
+Then delete the persistent `/var/lib/docker` directory, and restart the Docker daemon.
+
+If the file is empty, then you'd add it like:
+
+```
+EXTRA_ARGS="--usernc-remap=default
```
Comment (tianon) [inline doc/FAQ.md]: `s/usernc/userns/` :smile:

**Pull request closed without merging**